src/utils/actions.js

```diff
@@ -32,37 +35,46 @@ export function wrapInDispatch(
     meta,
     payload: isObject(requestingType) ? requestingType.payload : { args },
   });
-  return ref[method](...args)
-    .then(result => {
-      const successIsObject = isObject(successType);
-      // Built action object handling function for custom payload
-      const actionObj = {
-        type: successIsObject ? successType.type : successType,
-        meta,
-        payload:
-          successIsObject && successType.payload
-            ? makePayload(successType, result)
-            : { args },
-      };
-      // Attach preserve to action if it is passed
-      if (successIsObject && successType.preserve) {
-        actionObj.preserve = successType.preserve;
-      }
-      // Attach merge to action if it is passed
-      if (successIsObject && successType.merge) {
-        actionObj.merge = successType.merge;
-      }
-      dispatch(actionObj);
-      return result;
-    })
-    .catch(err => {
-      dispatch({
-        type: errorType,
-        meta,
-        payload: err,
-      });
-      return Promise.reject(err);
+  const success = (result) => {
+    const successIsObject = isObject(successType);
+    // Built action object handling function for custom payload
+    const actionObj = {
+      type: successIsObject ? successType.type : successType,
+      meta,
+      payload:
+        successIsObject && successType.payload
+          ? makePayload(successType, result)
+          : { args },
+    };
+    // Attach preserve to action if it is passed
+    if (successIsObject && successType.preserve) {
+      actionObj.preserve = successType.preserve;
+    }
+    // Attach merge to action if it is passed
+    if (successIsObject && successType.merge) {
+      actionObj.merge = successType.merge;
+    }
+    dispatch(actionObj);
+    return result;
+  };
+  const failure = (err) => {
+    dispatch({
+      type: errorType,
+      meta,
+      payload: err,
     });
+    return Promise.reject(err);
+  };
+
```

Comment:
If we dynamically pick our promise here, we could just have the success/failure afterwords:

```js
const actionPromise = method === 'mutate' ? mutate(ref, ...args) : ref[method](...args)
return actionPromise
  .then((result) => {
    const successIsObject = isObject(successType);
    // Built action object handling function for custom payload
    const actionObj = {
      type: successIsObject ? successType.type : successType,
      meta,
      payload:
        successIsObject && successType.payload
          ? makePayload(successType, result)
          : { args },
    };
    // Attach preserve to action if it is passed
    if (successIsObject && successType.preserve) {
      actionObj.preserve = successType.preserve;
    }
    // Attach merge to action if it is passed
    if (successIsObject && successType.merge) {
      actionObj.merge = successType.merge;
    }
    dispatch(actionObj);
    return result;
  })
  .catch((err) => {
    dispatch({
      type: errorType,
      meta,
      payload: err,
    });
    return Promise.reject(err);
  })
```

Reply:
Yep that's better. Fixed.